### PR TITLE
Allow nulls for write elements in MXSerializer

### PR DESCRIPTION
**Pull request by garydgregory**
https://issues.apache.org/jira/browse/MJAVADOC-793

Allow nulls for write elements in `MXSerializer`.

### CODE REVIEW

**Comment by slawekjaranowski** [comment]
@garydgregory thanks ... I afraid thet we will have a conflict with my PR #28 

If you let me I first merge my PR and that you can rebase your

**Comment by garydgregory** [comment]
@slawekjaranowski 
OK, done.

**Comment by slawekjaranowski** [comment]
@michael-o @elharo can you also look at it 😄

**Comment by michael-o** [comment]
I'd like to see this rebased first...

**Comment by garydgregory** [comment]
I rebased on master again.

File: src/test/java/org/codehaus/plexus/util/xml/pull/MXSerializerTest.java [hunk 1]
```diff
@@ -55,4 +56,17 @@ private String expectedOutput() {
...
+     * Tests MJAVADOC-793.
+     */
+    @Test
+    public void testMJAVADOC793() throws IOException {
+        // should be no-ops
+        new MXSerializer().writeElementContent(null, null);
+        new MXSerializer().writeAttributeValue(null, null);
+        final StringWriter stringWriter = new StringWriter();
+        new MXSerializer().writeElementContent(null, stringWriter);
+        new MXSerializer().writeAttributeValue(null, stringWriter);
```
Comment (slawekjaranowski) [inline src/test/java/org/codehaus/plexus/util/xml/pull/MXSerializerTest.java:70]: should support ?

```java
new MXSerializer().writeElementContent("value", null);
new MXSerializer().writeAttributeValue("value", null);
```

File: src/main/java/org/codehaus/plexus/util/xml/pull/MXSerializer.java [hunk 2]
```diff
@@ -854,110 +854,114 @@ public void flush() throws IOException {
...
-            // TODO: check if doing char[] text.getChars() would be faster than getCharAt(i) ...
-            char ch = text.charAt(i);
-            if (ch == ']') {
-                if (seenBracket) {
-                    seenBracketBracket = true;
-                } else {
-                    seenBracket = true;
-                }
-            } else {
+        if (value != null) {
```
Comment (slawekjaranowski) [inline src/main/java/org/codehaus/plexus/util/xml/pull/MXSerializer.java:857]: maybe 
```java
if (value == null) {
  return;
}
```
in such way rest of code not needed be changed

**Comment by slawekjaranowski** [comment]
Please also change of PR / commit message to show what really is changed, what is root cause eg

`Allow nulls for write elements in MXSerializer`

**Comment by michael-o** [comment]
Test:
```
	 public static void main(String[] args) throws ParserConfigurationException, TransformerException {
			Document doc = DocumentBuilderFactory.newInstance().newDocumentBuilder().newDocument();
			Element root = doc.createElement("root");
			root.setTextContent(null);
			root.setAttribute("foo", null);
			doc.appendChild(root);

			DOMSource domSource = new DOMSource(doc);
			StreamResult result = new StreamResult(System.out);
	        TransformerFactory tf = TransformerFactory.newInstance();
	        Transformer transformer = tf.newTransformer();
	        transformer.setOutputProperty(OutputKeys.INDENT, "yes");
	        transformer.setOutputProperty("{http://xml.apache.org/xslt}indent-amount", "2");
	        transformer.transform(domSource, result);
		}
```
output:
```
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<root foo=""/>
```

**Comment by michael-o** [comment]
What about the pretty printer?

File: src/test/java/org/codehaus/plexus/util/xml/pull/MXSerializerTest.java (hunk 1 above)
Comment (slawekjaranowski) [reply src/test/java/org/codehaus/plexus/util/xml/pull/MXSerializerTest.java:70]: Also add some of assertions for what content of `stringWriter` should be.

File: src/main/java/org/codehaus/plexus/util/xml/pull/MXSerializer.java [hunk 3]
```diff
@@ -854,110 +854,114 @@ public void flush() throws IOException {
...
+            // .[apostrophe and <, & escaped],
+            final char quot = attributeUseApostrophe ? '\'' : '"';
```
Comment (elharo) [inline src/main/java/org/codehaus/plexus/util/xml/pull/MXSerializer.java:859]: Not sure why we even make a choice here. Would be simpler to pick one and stick with it.

File: src/main/java/org/codehaus/plexus/util/xml/pull/MXSerializer.java [hunk 4]
```diff
@@ -854,110 +854,114 @@ public void flush() throws IOException {
...
                     pos = i + 1;
-                } else if (seenBracketBracket && ch == '>') {
-                    if (i > pos) out.write(text.substring(pos, i));
-                    out.write("&gt;");
+                } else if (ch == quot) {
+                    if (i > pos) out.write(value.substring(pos, i));
+                    out.write(quotEntity);
                     pos = i + 1;
                 } else if (ch < 32) {
                     // in XML 1.0 only legal character are #x9 | #xA | #xD
```
Comment (elharo) [inline src/main/java/org/codehaus/plexus/util/xml/pull/MXSerializer.java:879]: legal C0 characters

File: src/main/java/org/codehaus/plexus/util/xml/pull/MXSerializer.java [hunk 5]
```diff
@@ -854,110 +854,114 @@ public void flush() throws IOException {
...
-                    if (ch == 9 || ch == 10 || ch == 13) {
-                        // pass through
-
-                        // } else if(ch == 13) { //escape
+                    // and they must be escaped otherwise in attribute value they are normalized to spaces
```
Comment (elharo) [inline src/main/java/org/codehaus/plexus/util/xml/pull/MXSerializer.java:880]: escaped;

File: src/main/java/org/codehaus/plexus/util/xml/pull/MXSerializer.java [hunk 6]
```diff
@@ -854,110 +854,114 @@ public void flush() throws IOException {
...
+                    if (ch == 13 || ch == 10 || ch == 9) {
+                        if (i > pos) out.write(value.substring(pos, i));
+                        out.write("&#");
+                        out.write(Integer.toString(ch));
+                        out.write(';');
+                        pos = i + 1;
+                    } else {
+                        throw new IllegalStateException(
+                                "character " + Integer.toString(ch) + " is not allowed in output" + getLocation());
```
Comment (elharo) [inline src/main/java/org/codehaus/plexus/util/xml/pull/MXSerializer.java:889]: IllegalArgumentException

File: src/main/java/org/codehaus/plexus/util/xml/pull/MXSerializer.java [hunk 7]
```diff
@@ -854,110 +854,114 @@ public void flush() throws IOException {
...
+                        // in XML 1.1 legal are [#x1-#xD7FF]
```
Comment (elharo) [inline src/main/java/org/codehaus/plexus/util/xml/pull/MXSerializer.java:890]: I'd delete all this commented code

File: src/main/java/org/codehaus/plexus/util/xml/pull/MXSerializer.java [hunk 8]
```diff
@@ -854,110 +854,114 @@ public void flush() throws IOException {
...
+                    } else if (seenBracketBracket && ch == '>') {
+                        if (i > pos) out.write(text.substring(pos, i));
+                        out.write("&gt;");
+                        pos = i + 1;
+                    } else if (ch < 32) {
+                        // in XML 1.0 only legal character are #x9 | #xA | #xD
+                        if (ch == 9 || ch == 10 || ch == 13) {
+                            // pass through
+
+                            // } else if(ch == 13) { //escape
```
Comment (elharo) [inline src/main/java/org/codehaus/plexus/util/xml/pull/MXSerializer.java:943]: delete commented code

**Review by elharo — APPROVED** [review]
some suggestion. I don't see anything actually incorrect here.

File: src/test/java/org/codehaus/plexus/util/xml/pull/MXSerializerTest.java (hunk 1 above)
Comment (garydgregory) [reply src/test/java/org/codehaus/plexus/util/xml/pull/MXSerializerTest.java]: Done.

**Comment by garydgregory** [comment]
@elharo Thank you for your comments. I'll let the Maven team address those since I feel they are out of scope for this PR.

File: src/test/java/org/codehaus/plexus/util/xml/pull/MXSerializerTest.java [hunk 9]
```diff
@@ -55,4 +56,19 @@ private String expectedOutput() {
         out.append("</root>");
         return out.toString();
     }
+
+    /**
+     * Tests MJAVADOC-793.
+     */
+    @Test
+    public void testMJAVADOC793() throws IOException {
```
Comment (michael-o) [inline src/test/java/org/codehaus/plexus/util/xml/pull/MXSerializerTest.java:64]: This should be renamed. This is not related to the Javadoc plugin. IT simply does not allow null values. `testNullValues()`?

**Review by michael-o — CHANGES_REQUESTED** [review]

**Comment by slawekjaranowski** [comment]
> @elharo Thank you for your comments. I'll let the Maven team address those since I feel they are out of scope for this PR.

you right if needed should be fixed in separate PR

File: src/test/java/org/codehaus/plexus/util/xml/pull/MXSerializerTest.java (hunk 9 above)
Comment (garydgregory) [reply src/test/java/org/codehaus/plexus/util/xml/pull/MXSerializerTest.java]: Done: `testWriteNullValues()`.

**Review by slawekjaranowski — APPROVED** [review]

**Comment by slawekjaranowski** [comment]
@michael-o I hope your doubts was resolved, so I will merge and process forward

**Comment by garydgregory** [comment]
TY @slawekjaranowski

File: src/test/java/org/codehaus/plexus/util/xml/pull/MXSerializerTest.java [hunk 10]
```diff
@@ -55,4 +56,19 @@ private String expectedOutput() {
         out.append("</root>");
         return out.toString();
     }
+
+    /**
+     * Tests MJAVADOC-793.
```
Comment (michael-o) [inline src/test/java/org/codehaus/plexus/util/xml/pull/MXSerializerTest.java:61]: Well, that comment is wrong here as well.

**Comment by michael-o** [comment]
> @michael-o I hope your doubts was resolved, so I will merge and process forward

Yes, but my question about the `PrettyPrinter` remains open.
